modules/apps/commerce/headless/headless-commerce/headless-commerce-delivery-cart-impl/src/main/java/com/liferay/headless/commerce/delivery/cart/internal/resource/v1_0/CartTransitionResourceImpl.java

```diff
@@ -82,6 +98,41 @@ public Page<CartTransition> getCartCartTransitionsPage(Long cartId)
 			else if (commerceOrder.isDraft()) {
 				transitionOVPs.add(new ObjectValuePair<>(0L, "submit"));
 			}
+
+			List<CommerceOrderItem> commerceOrderItems =
+				commerceOrder.getCommerceOrderItems();
+
+			if ((commerceOrder.getBillingAddressId() > 0) &&
+				Validator.isNotNull(
+					commerceOrder.getCommercePaymentMethodKey()) &&
+				(commerceOrder.getShippingAddressId() > 0) &&
+				!commerceOrderItems.isEmpty() &&
+				_isShippable(commerceOrder, commerceOrderItems)) {
+
+				int deliveryCommerceTermEntriesCount =
+					_commerceTermEntryLocalService.
+						getDeliveryCommerceTermEntriesCount(
+							contextCompany.getCompanyId(),
+							commerceOrder.getCommerceOrderTypeId(),
+							_getCommerceShippingFixedOptionId(commerceOrder));
+				int paymentCommerceTermEntriesCount =
+					_commerceTermEntryLocalService.
+						getPaymentCommerceTermEntriesCount(
+							contextCompany.getCompanyId(),
+							commerceOrder.getCommerceOrderTypeId(),
+							_getCommercePaymentMethodGroupRelId(commerceOrder));
+
+				if (((deliveryCommerceTermEntriesCount == 0) ||
+					 ((deliveryCommerceTermEntriesCount > 0) &&
+					  (commerceOrder.getDeliveryCommerceTermEntryId() > 0))) &&
+					((paymentCommerceTermEntriesCount == 0) ||
+					 ((paymentCommerceTermEntriesCount > 0) &&
+					  (commerceOrder.getPaymentCommerceTermEntryId() > 0)))) {
+
+					transitionOVPs.add(
+						new ObjectValuePair<>(0L, "quick-checkout"));
+				}
+			}
```

Comment:
I think this way the quick-checkout is still tied to a shippable order, I think you should decouple the shippable check something like this:

```suggestion
			if (isOrderReadyForProcessing(commerceOrder, commerceOrderItems)) {

	int paymentTermCount = _commerceTermEntryLocalService
		.getPaymentCommerceTermEntriesCount(
			contextCompany.getCompanyId(),
			commerceOrder.getCommerceOrderTypeId(),
			_getCommercePaymentMethodGroupRelId(commerceOrder));

	boolean isPaymentTermValid = isTermValid(
		paymentTermCount, commerceOrder.getPaymentCommerceTermEntryId());

	// delivery terms are to check only for shippable order
	boolean isDeliveryTermValid = true; 
	if (_isShippable(commerceOrder, commerceOrderItems)) {
		int deliveryTermCount = _commerceTermEntryLocalService
			.getDeliveryCommerceTermEntriesCount(
				contextCompany.getCompanyId(),
				commerceOrder.getCommerceOrderTypeId(),
				_getCommerceShippingFixedOptionId(commerceOrder));

		isDeliveryTermValid = isTermValid(
			deliveryTermCount, commerceOrder.getDeliveryCommerceTermEntryId());
	}

	// check if terms are valid
	if (isDeliveryTermValid && isPaymentTermValid) {
		transitionOVPs.add(new ObjectValuePair<>(0L, "quick-checkout"));
	}
}

private boolean isOrderReadyForProcessing(
	CommerceOrder commerceOrder, List<CommerceOrderItem> commerceOrderItems) {

	return (commerceOrder.getBillingAddressId() > 0) &&
		Validator.isNotNull(commerceOrder.getCommercePaymentMethodKey()) &&
		(commerceOrder.getShippingAddressId() > 0) &&
		!commerceOrderItems.isEmpty();
}

private boolean isTermValid(int termCount, long termEntryId) {
	return (termCount == 0) || ((termCount > 0) && (termEntryId > 0));
}

```

Reply:
This way won't work if the items are shippable and no shipping address and shipping method is set.